### PR TITLE
fix(CreatePolicy): RHICOMPL-1676 finish, progress and error reporting

### PR DESCRIPTION
**Pull request by vkrizan**
* Clear the "This usually takes a minute or two." message after the
policy is created.
* Awaits assignment of all rules, sequentially (can be improved).
* Updates the progress.
* Adds error handling.
* Any errors on the Edit modal are absorbed and not shown (needs a general improvement).

Note for reviewers: changes are separated by commits.

Successful creation:
Screenshot_2021-04-13 SCAP policies - Compliance Red Hat Insights(3)

Error during the policy creation:
Screenshot_2021-04-14 SCAP policies - Compliance Red Hat Insights

Progress (with deliberate slowness):
https://user-images.githubusercontent.com/7695766/114585652-2973c080-9c84-11eb-95d8-444842fa6cfd.mp4

## Secure Coding Practices Checklist GitHub Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist
- [X] Input Validation
- [X] Output Encoding
- [X] Authentication and Password Management
- [X] Session Management
- [X] Access Control
- [X] Cryptographic Practices
- [X] Error Handling and Logging
- [X] Data Protection
- [X] Communication Security
- [X] System Configuration
- [X] Database Security
- [X] File Management
- [X] Memory Management
- [X] General Coding Practices

### CODE REVIEW

File: src/SmartComponents/EditPolicy/usePolicy.js [hunk 1]
```diff
@@ -28,8 +28,14 @@ const usePolicy = () => {
     const [associateSystems] = useMutation(ASSOCIATE_SYSTEMS_TO_PROFILES);
     const [associateRules] = useMutation(ASSOCIATE_RULES_TO_PROFILE);
 
-    return async (policy, updatedPolicy) => {
+    return async (policy, updatedPolicy, onProgress) => {
```
Comment (akofink) [inline src/SmartComponents/EditPolicy/usePolicy.js:67]: If you don't want to keep checking `onProgress &&`, we could define a default empty function:

```js
return async (policy, updatedPolicy, onProgress = () => {}) => {
```

File: src/SmartComponents/EditPolicy/usePolicy.js [hunk 2]
```diff
@@ -28,8 +28,14 @@ const usePolicy = () => {
...
+        const selectedRuleRefIds = updatedPolicy?.selectedRuleRefIds || [];
+        const expectedUpdates = 3 + selectedRuleRefIds.length;
```
Comment (akofink) [inline src/SmartComponents/EditPolicy/usePolicy.js:69]: Magic number - could we define a well named constant?

File: src/SmartComponents/EditPolicy/usePolicy.js [hunk 3]
```diff
@@ -48,12 +54,14 @@ const usePolicy = () => {
             let {
                 data: { createProfile: { profile: { id } } }
             } = await createProfile({ variables: { input: policyInput } });
+            onProgress && onProgress((++progress) / expectedUpdates);
```
Comment (akofink) [inline src/SmartComponents/EditPolicy/usePolicy.js:94]: This line could be its own local function to DRY it a bit.

File: src/SmartComponents/CreatePolicy/FinishedCreatePolicy.js [hunk 4]
```diff
@@ -41,7 +45,9 @@ const FinishedCreatePolicy = ({
             benchmarkId,
             hosts: systemIds.map((id) => ({ id })),
             selectedRuleRefIds
-        }).then(() => {
+        };
+
+        updatePolicy(null, newPolicy, onProgress).then(() => {
             setPercent(100);
```
Comment (akofink) [inline src/SmartComponents/CreatePolicy/FinishedCreatePolicy.js:51]: Is this even necessary anymore?

File: src/SmartComponents/EditPolicy/usePolicy.js [hunk 5]
```diff
@@ -21,12 +21,42 @@ const useCreateBusinessObjective = () => {
...
+    return async ({ id }, hosts) => {
+        const { data } = await associateSystems({
+            variables: { input: {
+                id,
+                systemIds: hosts.map((h) => (h.id))
+            } }
+        });
+        return data?.associateSystems?.profile;
+    };
+};
```
Comment (akofink) [inline src/SmartComponents/EditPolicy/usePolicy.js]: :clap: nice refactor, it's way better just by splitting these. Why not go ahead and split these into their own files? We could also do this after Summit or in another PR.

**Review by akofink — APPROVED** [review]
Works great, thanks @vkrizan! A few minor comments inline. One thing I noticed is that the user doesn't have any button to click during an error, they just have to close the wizard:

image

It looks like there's supposed to be a 'Back' button. I guess that would allow users to try again?

File: src/SmartComponents/EditPolicy/usePolicy.js (hunk 2 above)
Comment (vkrizan) [reply src/SmartComponents/EditPolicy/usePolicy.js:69]: Defining it outside of the function could lead to a mistake. We can detinitely create a named constant for it.

File: src/SmartComponents/EditPolicy/usePolicy.js (hunk 5 above)
Comment (vkrizan) [reply src/SmartComponents/EditPolicy/usePolicy.js]: Definitely. :wink:

File: src/SmartComponents/CreatePolicy/FinishedCreatePolicy.js (hunk 4 above)
Comment (vkrizan) [reply src/SmartComponents/CreatePolicy/FinishedCreatePolicy.js:51]: Playing it safe. :slightly_smiling_face:

**Comment by vkrizan** [comment]
@akofink Added the Back button.

I also had to add a spinner to the Edit Policy modal on the Save button, because due to the fix it takes much longer to save almost any change.

Please review once more. Thank you. :slightly_smiling_face:

File: src/SmartComponents/CreatePolicy/FinishedCreatePolicy.js [hunk 6]
```diff
@@ -57,6 +64,17 @@ const FinishedCreatePolicy = ({
         ));
     }
 
+    let secondaryActions;
+    if (percent === 100 || failed) {
+        secondaryActions = (
+            <Button
+                variant={'primary'}
+                onClick={() => { onWizardFinish(); }}>
+                { failed ? 'Back' : 'Return to application' }
```
Comment (akofink) [inline src/SmartComponents/CreatePolicy/FinishedCreatePolicy.js:73]: 'Back' makes me think I'll be returned to the previous wizard step and allowed to try the request again, but it closes the wizard. I think this is the correct behavior; perhaps we can merge as is and double check with UX?

**Review by akofink — APPROVED** [review]
Thanks! Works as expected.

File: src/SmartComponents/CreatePolicy/FinishedCreatePolicy.js (hunk 6 above)
Comment (vkrizan) [reply src/SmartComponents/CreatePolicy/FinishedCreatePolicy.js:73]: Yes, I agree. I can revert it to the "Return to application" even for failure.
